Configurable UI branding, keyboard list and hide-able columns through a new config.js file

anuragaw opened the pull request:
## Description
<!--- Describe your changes in detail -->
We want to support hiding table columns, specifically in metrics table, through config file so that users can make the relevant bits hidden as per their organization. Current work will support the metrics table but can be extended to any table with minimal work in future.

Config file will take the key of the metrics column from metrics.js file for the sake of minimal changes and simplicity of development.

**Problem**: The keyboard list in the UI is not consistent across views such as in the instance wizard and in the register template form. There is also no way to custom about url/text and doc title and help URL in the UI.
**Root Cause**: The list is hardcoded in the UI allowing no centralised configuration.
**Solution**: Introduce a new `config.js` file installed at the `/usr/share/cloudstackmanagement/webapp/config.js` location. The config.js allows configurable keyboard list, about url/text, doc title,  and help URL.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

## How Has This Been Tested?
Testing and further documentation in progress and will be updated with proper notes.
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document -->
Image with all columns showing up -
![unhiddencolumn](https://user-images.githubusercontent.com/43956255/55551406-56398880-56f8-11e9-8abc-57a4aa6a497c.png)

Change in config.js file -
![config](https://user-images.githubusercontent.com/43956255/55551432-66516800-56f8-11e9-88b2-23c514bf06df.png)

View after change in config file -
![hiddencolumn](https://user-images.githubusercontent.com/43956255/55551449-710bfd00-56f8-11e9-9251-94139a1d5bad.png)

Another Example with Hosts Metrics - 

![hosts_allvisible](https://user-images.githubusercontent.com/43956255/55552502-250e8780-56fb-11e9-94c1-bc66e1ec7ebe.png)

![hosts_config](https://user-images.githubusercontent.com/43956255/55552506-28a20e80-56fb-11e9-9799-f1f21b05714f.png)

![hosts_hidden](https://user-images.githubusercontent.com/43956255/55552515-2e97ef80-56fb-11e9-8426-d0da7627c44d.png)

rhtyd commented:
At some point, let's merge this with PR https://github.com/apache/cloudstack/pull/3217 otherwise this will lead to merge conflicts for related features extending/using config.js

anuragaw commented:
@blueorangutan package

blueorangutan commented:
@anuragaw a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2678

rhtyd commented:
Put some after-loaded method/js file that does something looping over sections minimizing changes to the listview/widget/library:
```
            $.each(cloudStack.sections, function(idx, section) {                   
                if (section.listView) {                                            
                    section.listView.preFilter = function(args) {                  
                        const original = section.listView.preFilter(args);         
                        return ['name'] + original;                                                                                                                     
                    };                                                             
                    console.log(Object.keys(section.listView.fields));             
                }                                                                  
            }); 
```

rhtyd commented:
Closed https://github.com/apache/cloudstack/pull/3217 in favour of this one

anuragaw commented on ui/scripts/ui/widgets/listView.js:
```diff
@@ -860,6 +860,8 @@
         if (groupableColumns) {
             $tr.addClass('groupable-header-columns').addClass('groupable-header');
             $.each(fields, function(key) {
```
This was broken in existing code and needed fixing. If an upper field is hidden we must not draw it's containing columns.

anuragaw commented:
@rhtyd I have updated the PR with the changes we discussed.

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2708

anuragaw commented on ui/config.js:
```diff
@@ -0,0 +1,39 @@
+cloudStackOptions = {
+    aboutText: "label.app.name",
+    aboutTitle: "label.about.app",
+    docTitle: "label.app.name",
```
Done

anuragaw commented on ui/config.js:
```diff
@@ -0,0 +1,39 @@
+
+// Define custom options configurable by admins for UI
+cloudStackOptions = {
+    aboutText: "label.app.name",
```
Done

anuragaw commented:
@PaulAngus - addressed all the comments in the latest commit. There are 5 keyboards specified in the config.js. I get first 4 of them. I wanted to confirm if the keyboard options you get were only missing "Simplified Chinese Keyboard" or any more?

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2765

rhtyd commented:
Purely UI changes, merging based on tests and reviews.

anuragaw commented:
Created PR https://github.com/apache/cloudstack/pull/3429 to clean up config file with redundant columns.
